File: leetcode/algorithms/415_add_strings/main.js
/**
 * @param {string} num1
 * @param {string} num2
 * @return {string}
 */
var addStrings = function (num1, num2) {
  while (num1.length != num2.length) {
    if (num1.length > num2.length) {
      num2 = "0" + num2;
    } else {
      num1 = "0" + num1;
    }
  }

  let result = "";
  let up = 0;

  for (let i = num1.length; i >= 0; i--) {
    const n1 = parseInt(num1[i]);
    const n2 = parseInt(num2[i]);
    let sum = n1 + n2 + up;
    up = 0;

    if (sum > 9) {
      up += 1;
      sum -= 10;
    }

    result += sum.toString();
  }

  return result;
};


// Solution
var solution = function (num1, num2) {
  let i = num1.length - 1;
  let j = num2.length - 1;
  let carry = 0;
  let sum = "";

  for (; i >= 0 || j >= 0 || carry > 0; i--, j--) {
    const digit1 = i < 0 ? 0 : num1.charAt(i) - "0";
    const digit2 = j < 0 ? 0 : num2.charAt(j) - "0";
    const digitsSum = digit1 + digit2 + carry;
    sum = `${digitsSum % 10}${sum}`;
    carry = Math.floor(digitsSum / 10);
  }

  return sum;
};

File: leetcode/algorithms/415_add_strings/main.java
class AddStrings {
    public String addStrings(String num1, String num2) {
        String result = "";
        int length = num1.length() > num2.length() ? num1.length() : num2.length();

        for (int i = 0; i < length; i++) {
            if (num1.length() < length) {
                num1 = "0" + num1;
            }
            if (num2.length() < length) {
                num2 = "0" + num2;
            }
        }

        int up = 0;
        for (int i = length - 1; i >= 0; i--) {
            int num = Integer.parseInt(String.valueOf(num1.charAt(i)))
                    + Integer.parseInt(String.valueOf(num2.charAt(i))) + up;
            up = 0;
            if (num > 9) {
                num -= 10;
                up += 1;
            }

            result = String.valueOf(num) + result;

            if (i == 0 && up > 0) {
                result = String.valueOf(up) + result;
            }
        }

        return result;
    }


    // Solution
    public String solution(String num1, String num2) {
        StringBuilder sb = new StringBuilder();
        int carry = 0;

        for (int i = num1.length() - 1, j = num2.length() - 1; i >= 0 || j >= 0 || carry == 1; i--, j--) {
            int x = i < 0 ? 0 : num1.charAt(i) - '0';
            int y = j < 0 ? 0 : num2.charAt(j) - '0';
            sb.append((x + y + carry) % 10);
            carry = (x + y + carry) / 10;
        }

        return sb.reverse().toString();
    }
}